### PR TITLE
gh-109054: Document configure variables

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-109054

----
:books: Documentation preview :books:: https://cpython-previews--109224.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
I'm not sure if adding a new dedicated ``Variables`` section is the best home for all these variables.

I chose to use ``.. cmdoption:: NAME`` syntax to avoid name conflicts with compiler flags defined with ``.. envvar:: NAME`` such as: https://docs.python.org/dev/using/configure.html#compiler-flags

I used ``./configure --help | grep '^\s\+[A-Z]\{2,\}'`` command to list all ``configure`` variables, see @erlend-aasland's comment: https://github.com/python/cpython/pull/109101#issuecomment-1711282393

Initially, I just wrote this PR to document ``LIBATOMIC`` :-)

File: Doc/using/configure.rst [hunk 1]
```diff
@@ -94,6 +94,154 @@ List all :file:`configure` script options using::
...
+
+   C compiler and linker flags for ``libpanel`` or ``libpanelw``, used by
+   :mod:`curses.panel` module, overriding ``pkg-config``.
+
+.. cmdoption:: PKG_CONFIG
+
+   Path to ``pkg-config`` utility.
+
+.. cmdoption:: PKG_CONFIG_LIBDIR
+.. cmdoption:: PKG_CONFIG_PATH
```
Comment (vstinner) [inline Doc/using/configure.rst]: I don't know PKG_CONFIG_LIBDIR and PKG_CONFIG_PATH.

File: Doc/using/configure.rst [hunk 2]
```diff
@@ -94,6 +94,154 @@ List all :file:`configure` script options using::
...
+
+   C compiler flags.
+
+.. cmdoption:: CPP
+
+   C preprocessor command.
+
+.. cmdoption:: CPPFLAGS
+
+   C preprocessor flags, e.g. ``-I<include dir>``.
```
Comment (vstinner) [inline Doc/using/configure.rst]: Maybe we should put BOLT, CC, CFLAGS, CPP and CPPFLAGS in a "Compiler flags" separated section?

File: Doc/using/configure.rst [hunk 3]
```diff
@@ -94,6 +94,154 @@ List all :file:`configure` script options using::
...
+   :mod:`curses` module, overriding ``pkg-config``.
+
+.. cmdoption:: GDBM_CFLAGS
+.. cmdoption:: GDBM_LIBS
+
+   C compiler and linker flags for ``gdbm``.
+
+.. cmdoption:: HOSTRUNNER
+
+   Program to run CPython for the host platform for cross-compilation.
```
Comment (vstinner) [inline Doc/using/configure.rst]: Maybe move this to the Cross-compiling section below?

File: Doc/using/configure.rst [hunk 4]
```diff
@@ -94,6 +94,154 @@ List all :file:`configure` script options using::
...
+
+.. cmdoption:: LDFLAGS
+
+   Linker flags, e.g. ``-L<library directory>``.
+
+   See also :option:`LIBS`.
```
Comment (vstinner) [inline Doc/using/configure.rst]: Maybe move LIBS, LDFLAGS and LIBATOMIC to a new Linker flags section?

File: Doc/using/configure.rst [hunk 5]
```diff
@@ -94,6 +94,154 @@ List all :file:`configure` script options using::
...
+
+.. cmdoption:: CPPFLAGS
+
+   C preprocessor flags, e.g. ``-I<include dir>``.
+
+.. cmdoption:: CURSES_CFLAGS
+.. cmdoption:: CURSES_LIBS
+
+   C compiler and linker flags for ``libncurses`` or ``libncursesw``, used by
+   :mod:`curses` module, overriding ``pkg-config``.
```
Comment (vstinner) [inline Doc/using/configure.rst]: Maybe move all NAME_{CFLAGS,LIBS} options to a new "Module flags" section? (Any idea for a better section name?)

**Comment by vstinner** [comment]
I reorganized the doc. I'm not sure between ``envvar`` and ``cmdoption``. The current doc uses ``envvar``.

cc @hugovk

**Comment by vstinner** [comment]
@erlend-aasland: it seems like the whole "3.3.4. Libraries Options" section is new in Python 3.11, all xxx_CFLAGS and xxx_LIBS, no?

**Comment by vstinner** [comment]
PROFILE_TASK was added in Python 3.8.

File: Doc/using/configure.rst [hunk 6]
```diff
@@ -775,6 +925,10 @@ the version of the cross compiled host Python.
       ac_cv_file__dev_ptmx=yes
       ac_cv_file__dev_ptc=no
 
+.. cmdoption:: HOSTRUNNER
+
+   Program to run CPython for the host platform for cross-compilation.
```
Comment (vstinner) [inline Doc/using/configure.rst:923]: Hum, it seems like it has been added to Python 3.11.

**Comment by erlend-aasland** [comment]
> @erlend-aasland: it seems like the whole "3.3.4. Libraries Options" section is new in Python 3.11, all xxx_CFLAGS and xxx_LIBS, no?

Yes, it came with the setup.py migration we did.

File: Doc/using/configure.rst [hunk 7]
```diff
@@ -289,6 +289,143 @@ General Options
...
+.. cmdoption:: LIBS
+
+   Libraries to pass to the linker, e.g. ``-l<library>``.
+
+.. cmdoption:: MACHDEP
+
+   Name for machine-dependent library files.
+
+
+Libraries Options
```
Comment (erlend-aasland) [inline Doc/using/configure.rst]: How about _Third-party dependencies_ or _Options for third-party dependencies_?

File: Doc/using/configure.rst [hunk 8]
```diff
@@ -289,6 +289,143 @@ General Options
...
+
+   Path to ``pkg-config`` utility.
+
+.. cmdoption:: PKG_CONFIG_LIBDIR
+.. cmdoption:: PKG_CONFIG_PATH
+
+   ``pkg-config`` options.
+
+
+C Compiler Options
```
Comment (erlend-aasland) [inline Doc/using/configure.rst]: Nitpick:

```suggestion
C compiler options
```

File: Doc/using/configure.rst [hunk 9]
```diff
@@ -289,6 +289,143 @@ General Options
...
+.. cmdoption:: CPP
+
+   C preprocessor command.
+
+.. cmdoption:: CPPFLAGS
+
+   C preprocessor flags, e.g. ``-I<include dir>``.
+
+
+Linker Options
```
Comment (erlend-aasland) [inline Doc/using/configure.rst]: Nitpick:

```suggestion
Linker options
```

File: Doc/using/configure.rst (hunk 8 above)
Comment (erlend-aasland) [reply Doc/using/configure.rst]: Hm, I see the rest of this file is deviating from the devguide recommendation for capitalisation. Perhaps we'd better stay with the current style.

File: Doc/using/configure.rst (hunk 7 above)
Comment (vstinner) [reply Doc/using/configure.rst]: Go for "Options for third-party dependencies".

File: Doc/using/configure.rst (hunk 8 above)
Comment (vstinner) [reply Doc/using/configure.rst]: I wrote the first version of this doc, and I have the bad habit of using French conventions. Instead, I prefer to capitalize all existing titles: I updated my PR for that.

File: Doc/using/configure.rst (hunk 9 above)
Comment (vstinner) [reply Doc/using/configure.rst]: ditto: I prefer to capitalize all titles.

**Comment by vstinner** [comment]
@erlend-aasland: I updated my PR, would you mind to review it again?

File: Doc/using/configure.rst (hunk 8 above)
Comment (erlend-aasland) [reply Doc/using/configure.rst]: > Perhaps we'd better stay with the current style.

Ah, the prevalent style _is_ sentence case :) So, the PR should be updated to use sentence case.

File: Doc/using/configure.rst [hunk 10]
```diff
@@ -289,6 +289,145 @@ General Options
...
+
+.. cmdoption:: LIBATOMIC
+
+   Linker flags when ``cpython/pyatomic.h`` header file is used.
+
+   Default: ``LIBATOMIC='-latomic'`` if ``libatomic`` is needed, or
+   ``LIBATOMIC=''`` otherwise (also the default when Python is cross-compiled).
+
+   .. versionadded:: 3.13
+
```
Comment (erlend-aasland) [inline Doc/using/configure.rst]: This is going away with #109344:

```suggestion
```

File: Doc/using/configure.rst (hunk 8 above)
Comment (hugovk) [reply Doc/using/configure.rst]: The [devguide](https://devguide.python.org/documentation/style-guide/#capitalization) says:

> In the Python documentation, the use of sentence case in section titles is preferable, but consistency within a unit is more important than following this rule. If you add a section to a chapter where most sections are in title case, you can either convert all titles to sentence case or use the dominant style in the new section title.

So: 

* Either use the prevalent style (sentence case here), or convert to sentence case. 
* Don't convert others to uppercase.

File: Doc/using/configure.rst (hunk 10 above)
Comment (vstinner) [reply Doc/using/configure.rst]: I understand that you want me to remove this variable in the doc, so I removed it.

File: Doc/using/configure.rst (hunk 8 above)
Comment (vstinner) [reply Doc/using/configure.rst]: I understand that the devguide is "The Law", but I wrote basically the whole documentation Doc/using/configure.rst. And I explained to you why I made a mistake. Now you want to use The Law against changing my own doc? I'm not sure that I get your rationale.

**Comment by vstinner** [comment]
Oh sorry for the noise, I tried a tool which messed up with my PR. I fixed my PR.

File: Doc/using/configure.rst [hunk 11]
```diff
@@ -289,6 +289,136 @@ General Options
...
+Linker Options
+--------------
+
+.. cmdoption:: LDFLAGS
+
+   Linker flags, e.g. ``-L<library directory>``.
+
+.. cmdoption:: LIBS
+
+   Libraries to pass to the linker, e.g. ``-l<library>``.
```
Comment (ezio-melotti) [inline Doc/using/configure.rst]:
```suggestion
.. cmdoption:: LDFLAGS

   Linker flags, e.g. :samp:`-L{library_directory}`.

.. cmdoption:: LIBS

   Libraries to pass to the linker, e.g. :samp:`-l{library}`.
```
Here you could use `:samp:` instead.

**Review by ezio-melotti — COMMENTED** [review]
I would keep the PR focused and handle capitalization issues in a separate PR.
New titles should follow the existing conventions, even if it doesn't match the preferred capitalization specified by the devguide.  A follow-up PR can fix it.

**Comment by vstinner** [comment]
> I would keep the PR focused and handle capitalization issues in a separate PR.

Hum, ok, I reverted these changes. I didn't expect that changing a few titles would be so controversial :-)

File: Doc/using/configure.rst [hunk 12]
```diff
@@ -289,6 +289,136 @@ General Options
...
+
+   C compiler flags.
+
+.. cmdoption:: CPP
+
+   C preprocessor command.
+
+.. cmdoption:: CPPFLAGS
+
+   C preprocessor flags, e.g. ``-I<include dir>``.
```
Comment (ezio-melotti) [inline Doc/using/configure.rst]:
```suggestion
   C preprocessor flags, e.g. :samp:`-I{include_dir}`.
```
I missed one.

**Comment by vstinner** [comment]
@erlend-aasland @hugovk @ezio-melotti: I rebased my PR on the main branch, squashed commits, I reverted title casing change (don't use Titlecase Anymore). Please review the updated PR.

@ezio-melotti: I applied your sugsestions to use ``:samp:``, thanks.

**Comment by hugovk** [comment]
Looks good, although I'll leave approval for someone who knows about these flags 👍

**Review by AA-Turner — APPROVED** [review]
Looks good Sphinx-wise. (But same caveat as Hugo wrt the flags)

A

File: Doc/using/configure.rst [hunk 13]
```diff
@@ -289,6 +289,136 @@ General Options
...
+.. cmdoption:: LIBSQLITE3_CFLAGS
+.. cmdoption:: LIBSQLITE3_LIBS
+
+   C compiler and linker flags for ``libsqlite3``, used by :mod:`sqlite3`
+   module, overriding ``pkg-config``.
+
+.. cmdoption:: LIBUUID_CFLAGS
+.. cmdoption:: LIBUUID_LIBS
+
+   C compiler and linker flags for ``liuuid``, used by :mod:`uuid` module,
```
Comment (ezio-melotti) [inline Doc/using/configure.rst]:
```suggestion
   C compiler and linker flags for ``libuuid``, used by :mod:`uuid` module,
```
I think this is missing a `b`.

**Review by ezio-melotti — APPROVED** [review]

**Comment by vstinner** [comment]
Thanks for the reviews. If someone has more remarks, feel free to propose a PR :-)

**Comment by erlend-aasland** [comment]
LGTM! Thanks for documenting these things.

**Pull request merged**